src/main/java/org/team1540/robot2022/commands/shooter/Shooter.java

```diff
@@ -139,4 +153,59 @@ public boolean isSpunUp() {
         return Math.abs(getClosedLoopError()) < SmartDashboard.getNumber("shooter/tuning/targetError", 0)
                 && Math.abs(getVelocityRPM(shooterMotorFront) + getVelocityRPM(shooterMotorRear)) > 200; // Make sure the shooter is moving
     }
+
+    /**
+     * Record a shot
+     *
+     * @param frontVelocity  front shooter flywheel RPM
+     * @param rearVelocity   rear shooter flywheel RPM
+     * @param targetDistance target distance in inches
+     * @param hood           is the hood up?
+     */
+    public void recordShot(double frontVelocity, double rearVelocity, double targetDistance, boolean hood) {
+        // If the last shot hasn't been saved yet, save it with an unknown value
+        // (setLastShotResult sets lastShot to null when it's done recording)
+        if (lastShot != null) {
+            setLastShotResult(ShotResult.UNKNOWN);
+        }
+
+        lastShot = new ShootingState(frontVelocity, rearVelocity, targetDistance, hood);
+    }
+
+    /**
+     * Record the result of the last shot and save it
+     *
+     * @param result shot result
+     */
+    public void setLastShotResult(ShotResult result) {
+        if (lastShot != null) {
+            lastShot.result = result;
+            saveShotToFile(lastShot);
+            lastShot = null;
+        }
+    }
+
+    /**
+     * Save a shot in the optimization table
+     */
+    private void saveShotToFile(ShootingState shot) {
+        String jsonl = String.format("{'match': %d, 'replay': %d, 'alliance': '%s', 'location': %d, 'matchSeconds': %f," +
+                        "'frontRPM': %f, 'rearRPM': %f, 'targetDistance': %f, 'hood': %b, 'result': '%s'}",
+                DriverStation.getMatchNumber(),
+                DriverStation.getReplayNumber(),
+                DriverStation.getAlliance().toString(),
+                DriverStation.getLocation(),
+                DriverStation.getMatchTime(),
+                shot.frontVelocity,
+                shot.rearVelocity,
+                shot.targetDistance,
+                shot.hood,
+                shot.result);
+
+        try {
+            Files.write(Paths.get("/optimization.jsonl"), jsonl.getBytes(), StandardOpenOption.APPEND);
+        } catch (IOException e) {
+            System.out.println("ERROR: While appending to optimization file: " + e);
```

Comment:
```suggestion
            DriverStation.reportError("ERROR: While appending to optimization file: " + e, e.getStackTrace());
```
I think using DriverStation.reportError will be more noticeable with its red text and all